Add cargo package firewall documentation (#197)
phylum-dev/documentation: docs/package_firewall/cargo.md

Reviewer: Since this is a new file, it should be added to `site/sidebars.js` so that it shows up in the sidebar.

# Cargo Package Firewall

> The Cargo package firewall is based on Cargo's sparse [Registry Index API].

Reviewer 2: Should it be "Package Firewall" throughout our documentation? (a proper noun phrase)

Author: I think package firewall is not required to be a proper noun. In headlines it's nicer to read this way but it makes perfect sense even without knowing anything about it so I think it's fine the way it is.

This also is already present in the other ecosystem's docs so I think it should be fine.

[Registry Index API]: https://doc.rust-lang.org/cargo/reference/registry-index.html

## Configuration

> All configuration options will require a Phylum API key, since Phylum's package
> firewall requires authentication. You can find out how to generate one in our

Reviewer 2: nit: I think since clauses are usually found at the beginning of a sentence.
```suggestion
Since Phylum's package firewall requires authentication, all configuration
options require a Phylum API key. You can find out how to generate one in our
```

Author: I'm not a native speaker, but I prefer the current version and it is already present this way in the other ecosystems. It's better to put the important stuff first, so people have an easier time skipping over irrelevant paragraphs.

[API Keys documentation].

In the following examples, all API keys will be represented as
`<PHYLUM_API_KEY>`, so make sure to replace them with your generated key.

Additionally, if the default [policy] is not sufficient, a group can be passed
to evaluate all packages against the group's policy. To do this, just replace
`<PHYLUM_GROUP>` with the desired group name. The supplied API key **must** have
access to this group.

If Phylum's default [policy] is sufficient, you can remove all instances of
`<PHYLUM_GROUP>`.

[API Keys documentation]: ../knowledge_base/api-keys.md#generate-an-api-key
[policy]: ../knowledge_base/policy.md

### cargo

To use Phylum's Cargo registry, it first needs to be added to a
[`.cargo/config.toml` file]. On UNIX you can use `~/.cargo/config.toml` to make
the registry available to all projects.

```toml
[registries]

> phylum = { index = "sparse+https://cargo.staging.phylum.io/" }

Reviewer 3: Should this be prod? i.e., `cargo.phylum.io`

Author: Yes it should, just copy/pasted it over and forgot to change it.

> phylum = { index = "sparse+https://cargo.staging.phylum.io/" }

Reviewer 2: Do we want to have the staging URL here?

```

This makes the Phylum registry available to all Cargo projects, however by
default Cargo will still use the official registry. To ensure all requests go

> through Phylum, you'll have to override the crates.io registry in the

Reviewer: Just a suggestion, but it might look and read more easily if "crates.io" is either turned into a link or put in backticks: `crates.io`. If the change is made here, then it should be made in the other spot it shows up a few lines down.

> [`.cargo/config.toml` file]. This can be a project's cargo config if you do not

Reviewer: ```suggestion
[`.cargo/config.toml` file]. This can be a project's Cargo config if you do not
```
This is one of two places where it wasn't capitalized.

wish to globally run every install through Phylum.

The easiest way to enable Phylum is by replacing the crates.io source with
Phylum's registry. This has the advantage of creating `Cargo.toml` and
`Cargo.lock` files that are identical to the ones created for the official
registry.

```toml
[source.crates-io]
replace-with = "phylum"
```

> Since replacing the cargo source prevents you from using the official registry,

Reviewer: ```suggestion
Since replacing the Cargo source prevents you from using the official registry,
```
This is one of two places where it wasn't capitalized.

this is unsuitable if certain packages should still be installable without going
through Phylum. In this case you can just set Phylum to be the default registry
instead:

```toml
[registry]
default = "phylum"
```

Since Phylum's registry requires authentication, you'll also need to configure a
credential provider in your [`.cargo/config.toml` file]:

```toml
[registry]
global-credential-providers = ["cargo:token"]
```

Once the file-based `cargo:token` credential provider is configured, you can add
your authentication details to the [`.cargo/credentials.toml` file].

> The Phylum group and API key first need to be converted to base 64:

Reviewer: ```suggestion
The Phylum group and API key first need to be converted to Base64:
```
`Base64` appears to be the way to reference it.
Reference: https://en.wikipedia.org/wiki/Base64

```text

> printf "<PHYLUM_GROU>:<PHYLUM_API_KEY>" | base64

Reviewer 2: ```suggestion
printf "<PHYLUM_GROUP>:<PHYLUM_API_KEY>" | base64
```

```

> ⚠️ **WARNING** ⚠️
>
> Do not accidentally save your token into your shell history.

> Once converted to base 64, you can add it as a basic authentication token to

Reviewer: ```suggestion
Once converted to Base64, you can add it as a basic authentication token to
```
`Base64` appears to be the way to reference it.
Reference: https://en.wikipedia.org/wiki/Base64

your credentials:

```toml
[registries.phylum]
token = "Basic <BASE64_OUTPUT>"
```

[`.cargo/config.toml` file]: https://doc.rust-lang.org/cargo/reference/config.html
[`.cargo/credentials.toml` file]: https://doc.rust-lang.org/nightly/cargo/reference/config.html#credentials

A blocked package will show up in `cargo` as missing:

```text
    Updating `phylum` index
error: no matching package named `libc` found
location searched: registry `phylum`
required by package `testing v0.1.0 (/tmp/testing)`
```

Alternatively, if a specific version constraint matches the blocked version:

```text
    Updating `phylum` index
error: failed to select a version for the requirement `libc = "=0.2.155"`
candidate versions found which didn't match: 0.2.158, 0.2.157, 0.2.156, ...
location searched: `phylum` index
required by package `testing v0.1.0 (/tmp/testing)`
```

If a verison range is accepted by the manifest, the package manager will

> automatically attempt using a version that passes Phylum's policy.

Reviewer 2: nit/optional: I really don't know what is grammatically correct here, but this sounds more correct to me.
```suggestion
automatically attempt to use a version that passes Phylum's policy.
```
